File: tools/sanitizer/utils/filter.py
#!/usr/bin/python

import re
from pathlib import Path


class Filter:

	def __init__(self, filters):
		self.filters = [Filter._compile(filterStr) for filterStr in filters]

	"""
	Compile a filter
	"""

	@staticmethod
	def _compile(filterStr):
		index = 0
		regexpr = r''
		while index < len(filterStr):
			c = filterStr[index]
			if c == "*":
				if index + 1 < len(filterStr) and filterStr[index + 1] == "*":
					regexpr += r'.*'
					index += 1
				else:
					regexpr += r'[^/]*'
			else:
				regexpr += re.escape(c)
			index += 1
		return re.compile(regexpr + r'$')

	"""
	Match a filter with a string
	"""

	def match(self, path: Path) -> bool:
		for regexpr in self.filters:
			if re.match(regexpr, path.as_posix()):
				return True
		return False

File: tools/sanitizer/utils/nodejs/filter.mjs
export default class Filter {
  constructor(filters) { this.filters = filters.map((filterStr) => Filter._compile(filterStr)); }

  static _escapeRegExp(string) { return string.replace(/[.*+\-?^${}()|[\]\\]/g, "\\$&"); }

  static _compile(filterStr) {
	let index = 0;
	let regexpr = "^";
	while (index < filterStr.length) {
	  const c = filterStr[index];
	  if (c == "*") {
		if (index + 1 < filterStr.length && filterStr[index + 1] == "*") {
		  regexpr += ".*";
		  ++index;
		} else {
		  regexpr += "[^/]*";
		}
	  } else {
		regexpr += Filter._escapeRegExp(c);
	  }
	  ++index;
	}
	return new RegExp(regexpr + "$");
  }

  match(path) {
	for (const regexpr of this.filters) {
	  if (path.match(regexpr)) {
		return true;
	  }
	}
	return false;
  }
}
